### PR TITLE
wip: add sort grad

### PR DESCRIPTION
**Pull request by polvalente**

### CODE REVIEW

File: nx/lib/nx/defn/grad.ex [hunk 1]
```diff
@@ -440,6 +440,77 @@ defmodule Nx.Defn.Grad do
...
+    # iotas = []
+    # for dim, size in enumerate(shape):
+    #   dtype = np.int32 if size < np.iinfo(np.int32).max else np.int64
+    #   iotas.append(broadcasted_iota(dtype, shape, dim))
+    # sorted_tensor = sort_p.bind(*(tensor + (iotas[dimension],)), dimension=dimension,
+    #                       is_stable=is_stable, num_keys=num_keys)
+    # idx = tuple(sorted_tensor[-1] if i == dimension else iotas[i]
+    #             for i in range(len(shape)))
+    # g_out = tuple(t if type(t) is ad_util.Zero else t[idx] for t in g)
+    # return tuple(sorted_tensor[:-1]), g_out
```
Comment (polvalente) [inline nx/lib/nx/defn/grad.ex:454]: This is the code I tried to translate from Jax.
It seems that the basic idea is to retrieve the corresponding grads for each "line" of the sorted tensor, by logical indexing.

However, we don't have logical indexing available, so I tried another approach by concatenating a sequence of numbers to the tensor, slicing it afterwards. However, I couldn't find a way to convert the resulting {n} tensor into a list of integers, which in turn makes the construction of `starts` impossible.

Any tips on how we could approach this?

Comment (josevalim) [reply nx/lib/nx/defn/grad.ex:454]: What is logical indexing in this case? Are you selecting only some lines and ignoring the other ones? Or are you re-ordering?

Comment (polvalente) [reply nx/lib/nx/defn/grad.ex:454]: From what I gathered from the code, it basically reorders the gradient tensor according to the ordering applied to the original tensor

**Comment by polvalente** [comment]
I'm closing this PR for now because it'll be a lot easier to implement this grad when #223 or #221 are available

**Pull request closed without merging**